runtime/parachains/src/disputes.rs

```diff
@@ -717,82 +840,52 @@ impl<T: Config> Pallet<T> {
 	/// Handle sets of dispute statements corresponding to 0 or more candidates.
 	/// Returns a vector of freshly created disputes.
 	///
+	/// Assumes `statement_sets` were already de-duplicated.
+	///
 	/// # Warning
 	///
 	/// This functions modifies the state when failing. It is expected to be called in inherent,
 	/// and to fail the extrinsic on error. As invalid inherents are not allowed, the dirty state
 	/// is not committed.
-	pub(crate) fn provide_multi_dispute_data(
-		statement_sets: MultiDisputeStatementSet,
+	pub(crate) fn process_checked_multi_dispute_data(
+		statement_sets: CheckedMultiDisputeStatementSet,
 	) -> Result<Vec<(SessionIndex, CandidateHash)>, DispatchError> {
 		let config = <configuration::Pallet<T>>::config();
 
-		// Deduplicate.
-		{
-			let mut targets: Vec<_> =
-				statement_sets.iter().map(|set| (set.candidate_hash.0, set.session)).collect();
-
-			targets.sort();
-
-			let submitted = targets.len();
-			targets.dedup();
-
-			ensure!(submitted == targets.len(), Error::<T>::DuplicateDisputeStatementSets);
-		}
-
 		let mut fresh = Vec::with_capacity(statement_sets.len());
 		for statement_set in statement_sets {
-			let dispute_target = (statement_set.session, statement_set.candidate_hash);
-			if Self::provide_dispute_data(&config, statement_set)? {
+			let dispute_target = {
+				let statement_set: &DisputeStatementSet = statement_set.as_ref();
+				(statement_set.session, statement_set.candidate_hash)
+			};
+			if Self::process_checked_dispute_data(
+				statement_set,
+				config.dispute_post_conclusion_acceptance_period,
+			)? {
 				fresh.push(dispute_target);
 			}
 		}
 
 		Ok(fresh)
 	}
 
-	/// Removes all duplicate disputes.
-	fn filter_multi_dispute_data(statement_sets: &mut MultiDisputeStatementSet) {
-		frame_support::storage::with_transaction(|| {
-			let config = <configuration::Pallet<T>>::config();
-
-			let old_statement_sets = sp_std::mem::take(statement_sets);
-
-			// Deduplicate.
-			let dedup_iter = {
-				let mut targets = BTreeSet::new();
-				old_statement_sets.into_iter().filter(move |set| {
-					let target = (set.candidate_hash, set.session);
-					targets.insert(target)
-				})
-			};
-
-			*statement_sets = dedup_iter
-				.filter_map(|set| {
-					let filter = Self::filter_dispute_data(&config, &set);
-
-					filter.filter_statement_set(set)
-				})
-				.collect();
-			TransactionOutcome::Rollback(())
-		})
-	}
-
 	// Given a statement set, this produces a filter to be applied to the statement set.
 	// It either removes the entire dispute statement set or some specific votes from it.
 	//
 	// Votes which are duplicate or already known by the chain are filtered out.
 	// The entire set is removed if the dispute is ancient or concluded.
```

Comment:
Are you saying it should only filter if a statement set is both, ancient AND concluded, and only in that case it should be filtered? I don't think that's desired.

The `post_conclusion_acceptance_period` makes sure to handle late dispute votes as well. `ancient` already implies this.

Reply:
Right now the comment says, that we will remove a statement set if the dispute is concluded, but the code says otherwise: We only remove it, if it is concluded and older than `post_conclusion_acceptance_period` ... so we do accept late votes.

We also accept votes for "ancient" disputes, if they are not concluded yet. The only time we actually don't accept a statement set if the dispute is concluded and "ancient" ... meaning `post_conclusion_acceptance_period` has passed.

Reply:
Going to adjust the comment, thanks!